src/v/kafka/server/partition_proxy.cc

```diff
@@ -18,16 +18,21 @@
 namespace kafka {
 
 std::optional<partition_proxy> make_partition_proxy(
-  const model::materialized_ntp& mntp,
-  ss::lw_shared_ptr<cluster::partition> partition,
+  const model::ntp& ntp,
+  cluster::metadata_cache& md_cache,
   cluster::partition_manager& pm) {
-    if (!mntp.is_materialized()) {
-        return make_partition_proxy<replicated_partition>(partition);
+    auto log = pm.get(ntp);
+    if (log) {
+        return make_partition_proxy<replicated_partition>(log);
     }
-    if (auto log = pm.log(mntp.input_ntp()); log) {
-        return make_partition_proxy<materialized_partition>(*log);
+    auto mts = md_cache.get_source_topic(model::topic_namespace_view{ntp});
+    if (!mts) {
+        return std::nullopt;
     }
-    return std::nullopt;
+    auto source_ntp = model::ntp(ntp.ns, *mts, ntp.tp.partition);
+    auto materialized_log = pm.log(ntp);
+    vassert(materialized_log, "log should exist for materialized partition");
```

Comment:
I agree, lets remove both assertions, and return nullopt. Client will receive `no_topic_or_partition_exists` and it would be technically true if the underlying subsystems don't yet reflect what they are intended to contain.